### DISCUSSION TITLE
Automerge stopped working on GitLab Enterprise Edition 16.0.0-pre ab69edc3ede

### DISCUSSION THREAD

**Discussion by bourquep** [renovatebot/renovate#22063] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.75.0

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

gitlab.com Premium 16.0.0-pre ab69edc3ede

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Wanted end result.

Auto-merge MRs should auto-merge.

### What you tried so far.

Since this weekend (last successful auto-merge MR was May 5th), all our auto-merge MRs on Gitlab started not auto-merging. What I observe:

* In the configuration report of the MR, I see that auto merge is enabled: ![CleanShot 2023-05-09 at 19 54 56@2x](https://github.com/renovatebot/renovate/assets/65096/5127d31b-9618-4a4d-ba65-b1baad0fea17)
* There are no reviewers assigned to this MR, as expected for an auto merge MR
* The MR's pipeline succeeds
* I _do not_ see the typical _StudyoBot enabled an automatic merge when the pipeline for SHA succeeds_ mention in the MR's activity log.
* Renovate's logs are attached below.

So when Renovate call Gitlab's API to set the number of required approvers to zero, the API returns `400` with a `any-approver for the merge request already exists` message.

As far as I can tell, [this code](https://github.com/renovatebot/renovate/blob/0755daa4818ed05e151afe8d35b3c43bd22529ca/lib/modules/platform/gitlab/index.ts#L537-L542) is responsible for making that `POST` request.

It looks as if Gitlab is refusing to create the `renovateIgnoreApprovals` rule because a system-defined rule for _zero approvers_ already exists (aka `any_approver`, see [this doc](https://docs.gitlab.com/ee/api/merge_request_approvals.html#create-project-level-rule) for a mention of that system rule).

I _feel_ like this might be a breaking change introduced by Gitlab 16, which will be released to all at the end of May but which I seem to be using (my gitlab.com reported version is `GitLab Enterprise Edition 16.0.0-pre ab69edc3ede`).

Looking at the breaking changes of v16, I see [this mention](https://docs.gitlab.com/ee/update/deprecations.html?removal_milestone=16.0&breaking_only=true#changing-merge-request-approvals-with-the-approvals-api-endpoint) of a change in the `approvals` endpoint. That breaking change is not directly related to the issue at hand, but it is _possible_ that the _return `400` when creating a zero-approver rule_ breaking change might have been introduced at the same time.

I believe that the [current Renovate code](https://github.com/renovatebot/renovate/blob/0755daa4818ed05e151afe8d35b3c43bd22529ca/lib/modules/platform/gitlab/index.ts#L530-L547) that handles setting a zero-approver rule to a MR should be updated, but I'm not sure that I am qualified to do this myself.

~I _believe_ that the code should be changed to perform a simple `PUT`, passing `any_approver` as `:approval_rule_id`. See [this doc](https://docs.gitlab.com/ee/api/merge_request_approvals.html#update-merge-request-level-rule).~

**Edit**: I have tried calling the `PUT /projects/:id/merge_requests/:merge_request_iid/approval_rules/:approval_rule_id` API with `any_approver` or `any-approver` as the `approval_rule_id` and it failed.

I have also tried calling `POST /projects/:id/merge_requests/:merge_request_iid/approval_rules` (just as Renovate does) and it does indeed return `400`:

```
POST /api/v4/projects/studyoco%2Ftoday%2Ftoday-web/merge_requests/991/approval_rules HTTP/1.1
Accept: application/json, */*;q=0.5
Accept-Encoding: gzip, deflate
Authorization: Bearer glpat-*******
Connection: keep-alive
Content-Length: 45
Content-Type: application/json
Host: gitlab.com
User-Agent: HTTPie/3.2.1

{
    "approvals_required": "0",
    "name": "foobar"
}


HTTP/1.1 400 Bad Request

{
    "message": {
        "rule_type": [
            "any-approver for the merge request already exists"
        ]
    }
}
```

I now believe this is a bug with Gitlab, most probably introduced by v16. I have opened a support ticket with Gitlab, will post updates here as soon as I have a reply from them.

### Relevant debug logs

<details><summary>Logs</summary>

```
INFO: Repository started (repository=studyoco/today/today-web)
       "renovateVersion": "35.75.0"
 INFO: Dependency extraction complete (repository=studyoco/today/today-web, baseBranch=master)
       "stats": {
         "managers": {
           "bundler": {"fileCount": 1, "depCount": 1},
           "dockerfile": {"fileCount": 2, "depCount": 3},
           "gitlabci": {"fileCount": 3, "depCount": 8},
           "gradle": {"fileCount": 7, "depCount": 9},
           "gradle-wrapper": {"fileCount": 1, "depCount": 1},
           "jsonnet-bundler": {"fileCount": 1, "depCount": 2},
           "npm": {"fileCount": 1, "depCount": 129}
         },
         "total": {"fileCount": 16, "depCount": 153}
       }
 INFO: Branch created (repository=studyoco/today/today-web, branch=renovate/eslint-plugin-jsdoc-44.x)
       "commitSha": "9fa8a895c096add4b67f2cecee93edbb540953fb"
 WARN: GitLab: Error adding approval rule (repository=studyoco/today/today-web, branch=renovate/eslint-plugin-jsdoc-44.x)
       "err": {
         "name": "HTTPError",
         "code": "ERR_NON_2XX_3XX_RESPONSE",
         "timings": {
           "start": 1683673832928,
           "socket": 1683673832928,
           "lookup": 1683673832930,
           "connect": 1683673832938,
           "secureConnect": 1683673832949,
           "upload": 1683673832949,
           "response": 1683673833160,
           "end": 1683673833160,
           "phases": {
             "wait": 0,
             "dns": 2,
             "tcp": 8,
             "tls": 11,
             "request": 0,
             "firstByte": 211,
             "download": 0,
             "total": 232
           }
         },
         "message": "Response code 400 (Bad Request)",
         "stack": "HTTPError: Response code 400 (Bad Request)\n    at Request.<anonymous> (/opt/buildpack/tools/renovate/35.75.0/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/35.75.0 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "authorization": "***********",
             "content-type": "application/json",
             "content-length": "57",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://gitlab.com/api/v4/projects/studyoco%2Ftoday%2Ftoday-web/merge_requests/991/approval_rules",
           "hostType": "gitlab",
           "username": "",
           "password": "",
           "method": "POST",
           "http2": false
         },
         "response": {
           "statusCode": 400,
           "statusMessage": "Bad Request",
           "body": {
             "message": {
               "rule_type": ["any-approver for the merge request already exists"]
             }
           },
           "headers": {
             "date": "Tue, 09 May 2023 23:10:33 GMT",
             "content-type": "application/json",
             "content-length": "79",
             "connection": "close",
             "cache-control": "no-cache",
             "content-security-policy": "default-src 'none'",
             "vary": "Origin",
             "x-content-type-options": "nosniff",
             "x-frame-options": "SAMEORIGIN",
             "x-request-id": "aee6684114b8d05df9351ea814441d61",
             "x-runtime": "0.077787",
             "strict-transport-security": "max-age=31536000",
             "referrer-policy": "strict-origin-when-cross-origin",
             "ratelimit-observed": "34",
             "ratelimit-remaining": "1966",
             "ratelimit-reset": "1683673893",
             "ratelimit-resettime": "Tue, 09 May 2023 23:11:33 GMT",
             "ratelimit-limit": "2000",
             "gitlab-lb": "fe-16-lb-gprd",
             "gitlab-sv": "localhost",
             "cf-cache-status": "DYNAMIC",
             "report-to": "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=mxrZVx5XN8ja7CYqLfD05BNd%2FGbgeRi2O%2F4bgwN7GEC2dg%2BSVBda0TgHP0Lv1vAohK1ukiHi135JgcoIteJnnuvDPuGcoi9C1HTXpX1YfEHQ0QZNtdpIukp2d9Y%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}",
             "nel": "{\"success_fraction\":0.01,\"report_to\":\"cf-nel\",\"max_age\":604800}",
             "set-cookie": [
               "_cfuvid=hAdMs8Npq95zStVpIFxJzASSQP7OjTFYYSrtMp.LZbU-1683673833156-0-604800000; path=/; domain=.gitlab.com; HttpOnly; Secure; SameSite=None"
             ],
             "server": "cloudflare",
             "cf-ray": "7c4d9dcffc07a23b-YYZ"
           },
           "httpVersion": "1.1",
           "retryCount": 0
         }
       }
 INFO: PR created (repository=studyoco/today/today-web, branch=renovate/eslint-plugin-jsdoc-44.x)
       "pr": 991,
       "prTitle": "chore(deps): update dependency eslint-plugin-jsdoc to v44.1.0"
 INFO: Repository finished (repository=studyoco/today/today-web)
       "cloned": true,
       "durationMs": 66870
```

</details>

**Comment by bourquep** [comment]
Here's the reply I got from Gitlab Support:

> GitLab returns the "any-approver for the merge request already exists" error response when an any-approver rule already exists - you're receiving this due to an [MR Approval rule already being configured on the project that you've indicated](https://gitlab.com/studyoco/today/today-web/-/settings/merge_requests).
> 
> When a new Merge Request is created, the project-level approval rules are copied to the Merge Request - so since there is already a project-level Approval Rule configured, the MR already has an approval rule when you attempt to create a new one with your POST request.
> 
> To get around this in the short term, you can [update the merge request-level approval rule with a PUT request](https://docs.gitlab.com/ee/api/merge_request_approvals.html#update-merge-request-level-rule), however I understand since your bot will run on multiple MRs this isn't really a suitable solution.
> 
> To get around this there are a few options:
> 
> 1. You could remove the project-level approval rule - I wouldn't really recommend this as it would mean that the rule would not be applied to other project members.
> 2. You could add the bot as an approver on the [MR settings page](https://gitlab.com/studyoco/today/today-web/-/settings/merge_requests) - this would be the solution that I'd recommend.
> 
> There were a [few changes](https://about.gitlab.com/upcoming-releases/) in 16.0 relating to Merge Requests, but i'm not quite sure this has resulted in this particular break - I'll investigate internally with our team.

I don't believe his 2nd suggestion (add Renovate as an approver) would help, as Renovate doesn't have an _approve this MR_ step that I could find.

**Reply by rarkins** [reply]
I don't think I fully understand the problem, but would it work if Renovate simply swallowed/ignored that "any-approver" response and continued as if it succeeded? Is "any-approver" a GitLab keyword and consistent across every user, or is it a name for a rule you have chosen?

**Reply by bourquep** [reply]
AFAICT, it already ignores that error. Problem is that if the `ignoreApprovals` method fails, the number of required approvers on the MR will remain >0, which will prevent Renovate from auto-merging the MR.

**Reply by bourquep** [reply]
@rarkins I'm starting to wonder if it wouldn't be better for Renovate to _approve_ its own MRs when auto merge is true, rather than try to eliminate the _number of approvals_ rule on the MR...

**Reply by viceice** [reply]
are you sure a user can approve its own MR? every platform I've worked with would block that.

**Reply by bourquep** [reply]
It's a setting in each repo. But yes, we can't assume that self-approvals are allowed.

**Reply by bourquep** [reply]
I received further details from Gitlab Support:

> Previous to version 16 MRs were being created via the API without any approval rules altogether, which was amended in [this merge request](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/118235). Please note that this problem was limited to the API - MRs created via the UI were created with approval rules inherited from the project-scope as intended.
>
> As a result of this MR, merge requests created via the API will be populated with the approval rules from the project-scope. I understand this is likely frustrating given the bot appears to have been written under the understanding that those rules are not inherited.

The support agent is proposing the following 2 options:

> 1. The bot could be modified to create the MR with the rules as part of the creation payload. This could utilise the approvals_before_merge optional parameter [[doc]](https://docs.gitlab.com/ee/api/merge_requests.html#create-mr) set to 0.
> 2. The bot could update the project-level rule after MR creation using [the PUT endpoint](https://docs.gitlab.com/ee/api/merge_request_approvals.html#update-merge-request-level-rule). This could be implemented [in the function you indicated](https://github.com/renovatebot/renovate/blob/0755daa4818ed05e151afe8d35b3c43bd22529ca/lib/modules/platform/gitlab/index.ts#L530-L547) in your previous message

I have experimented with option 2, and it works. But it requires to first do a `GET` call to obtain the list of rules applied to the MR:

```
http GET https://gitlab.com/api/v4/projects/studyoco%2Ftoday%2Ftoday-web/merge_requests/991/approval_rules Authorization:"Bearer glpat-******"
HTTP/1.1 200 OK

[
    {
        "approvals_required": 1,
        "contains_hidden_groups": false,
        "eligible_approvers": [],
        "groups": [],
        "id": 93941727,
        "name": "All Members",
        "overridden": false,
        "rule_type": "any_approver",
        "section": null,
        "source_rule": {
            "approvals_required": 1
        },
        "users": []
    }
]
```

Then read the `id` of the rule with `rule_type: any_approver`, and `PUT` that rule ID on the MR, with `approvals_required` set to `0`:

```
http PUT https://gitlab.com/api/v4/projects/studyoco%2Ftoday%2Ftoday-web/merge_requests/991/approval_rules/93941727 approvals_required=0 Authorization:"Bearer glpat-******"
HTTP/1.1 200 OK

{
    "approvals_required": 0,
    "contains_hidden_groups": false,
    "eligible_approvers": [],
    "groups": [],
    "id": 93941727,
    "name": "All Members",
    "overridden": true,
    "rule_type": "any_approver",
    "section": null,
    "source_rule": {
        "approvals_required": 1
    },
    "users": []
}
```

I suppose that if the `GET` call returns nothing, we should then create our own rule in the same way as is currently done by Renovate.

Option 1 might be more straightforward though, as it only requires setting the `approvals_before_merge` to `0` when creating the MR.

**Reply by rarkins** [reply]
A question is whether we need to make this configurable if either:
- Users may not want this, or
- It can cause errors and needs turning off

If neither is the case and it's "safe" to do for every platform automerge MR, then we shouldn't need a new config option

**Reply by bourquep** [reply]
I believe that the existing config option [`gitLabIgnoreApprovals`](https://github.com/renovatebot/renovate/blob/main/docs/usage/configuration-options.md#gitlabignoreapprovals) is sufficient as it is already documented as _so that Renovate bot can automerge its merge requests, without needing approval(s)_

Both proposed options would still honour that definition.

**Reply by cjpearson** [reply]
Hey, I'm experiencing this issue as well. Option 1 sounds better to me since I think it would handle more scenarios. For example, if your project has multiple default approval rules or requires a specific reviewer, I would expect `gitLabIgnoreApprovals` to clear them all.

Perhaps another approach would be to delete the existing rules after creation? Looking at [this code](https://github.com/renovatebot/renovate/blob/bf2f9db59d87f0b07290d4f35ce169c3eb494328/lib/modules/platform/gitlab/index.ts#L530), it first fetches all the rules for the MR then creates the renovate rule if it does not exist. I suppose this could be extended to [delete](https://docs.gitlab.com/ee/api/merge_request_approvals.html#delete-merge-request-level-rule) each of the non-renovate rules?

**Reply by rarkins** [reply]
I think you're right that the existing `gitLabIgnoreApprovals` is explicit enough in its intent that Renovate can do more things to try to ignore approvals and get the MR automerged. But we should aim for a way which has least side effects and is resilient towards subsequent pushes (e.g. branch needs updating or de-conflicting after its initial creation).

PRs welcome which satisfy the above

**Reply by bourquep** [reply]
Gitlab Support just mentioned that option 1 _will not work as the approvals_before_merge value must be at least equal to the project-level approvals_before_merge value (which in this case is 1 - sorry!)._

**Reply by a1flecke** [reply]
@rarkins 

I looked at the merge approval rules json coming back from Gitlab:

```json
[{"id":94192565,"name":"All Members","rule_type":"any_approver","eligible_approvers":[],"approvals_required":1,"users":[],"groups":[],"contains_hidden_groups":false,"section":null,"source_rule":{"approvals_required":1},"overridden":false}]
```

looking at the error that gitlab returns, wouldn't a solution look like updating [this logic](https://github.com/renovatebot/renovate/blob/main/lib/modules/platform/gitlab/index.ts#L530-L546) to:

1. first check if a rule of type `any_approver` exists
2. if it does: do an [update](https://docs.gitlab.com/ee/api/merge_request_approvals.html#update-merge-request-level-rule) to set the `approvals_required` to 0
3. else if not: check if a renovate rule exists
4. if not: create
5. else: do nothing

Thoughts?

**Reply by bourquep** [reply]
@a1flecke That's exactly what I had in mind. I believe it would be the best way to fix this.

**Reply by rarkins** [reply]
If you set approvals_required to 1, is it Renovate which then approves its own?

**Reply by bourquep** [reply]
We should set `approvals_required` to `0`, not `1`

**Reply by fredrondina** [reply]
@bourquep @a1flecke We are also seeing this issue, and this looks like a good solution. Are we interested in trying this? I have some free time and could get a PR out.

**Reply by a1flecke** [reply]
@rarkins  @bourquep  thanks for catching the typo. I updated my comment to specify that it should be `0` and not `1`.  Its what I get for commenting while distracted

**Reply by bourquep** [reply]
@fredrondina I am definitely interested in trying this. If you have some free time, go for it! Thanks! 🙏

**Reply by rarkins** [reply]
Is this now the same as #5709?

**Reply by bourquep** [reply]
@rarkins AFAICT, #5709 is already implemented; this thread is about fixing the feature after Gitlab 16 introduced an API breaking change.

**Reply by rarkins** [reply]
What about https://github.com/renovatebot/renovate/issues/6473 or https://github.com/renovatebot/renovate/issues/11106 ?

**Reply by bourquep** [reply]
#6473 is a different use-case. Unsure about #11106

**Reply by fredrondina** [reply]
https://github.com/renovatebot/renovate/pull/22237

**Comment by bourquep** [comment, marked as answer]
For those following this issue, @fredrondina's PR has been merged and the fix is available in Renovate 35.87.2.

~The release hasn't yet made its way into the [Gitlab runner](https://gitlab.com/renovate-bot/renovate-runner/-/releases), but it should be picked up soon.~

[Gitlab Runner v13.8.2](https://gitlab.com/renovate-bot/renovate-runner/-/releases/v13.8.2) includes the fix.

Thanks to all who contributed to the discussion, and thanks to @fredrondina for the PR!
